lib/Minz/Migrator.php

```diff
@@ -0,0 +1,281 @@
+<?php
+
+/**
+ * The Minz_Migrator helps to migrate data (in a database or not) or the
+ * architecture of a Minz application.
+ *
+ * @author Marien Fressinaud <dev@marienfressinaud.fr>
+ * @license http://www.gnu.org/licenses/agpl-3.0.en.html AGPL
+ */
+class Minz_Migrator
+{
+	/** @var string[] */
+	private $applied_versions;
+
+	/** @var array */
+	private $migrations = [];
+
+	/**
+	 * Execute a list of migrations, skipping versions indicated in a file
+	 *
+	 * @param string $migrations_path
+	 * @param string $applied_migrations_path
+	 *
+	 * @throws BadFunctionCallException if a callback isn't callable.
+	 * @throws DomainException if there is no migrations corresponding to the
+	 *                         given version (can happen if version file has
+	 *                         been modified, or migrations path cannot be
+	 *                         read).
+	 *
+	 * @return boolean|string Returns true if execute succeeds to apply
+	 *                        migrations, or a string if it fails.
+	 */
+	public static function execute($migrations_path, $applied_migrations_path) {
+		if (!file_exists($applied_migrations_path)) {
+			return "{$applied_migrations_path} file does not exist";
+		}
+
+		$applied_migrations_file = fopen($applied_migrations_path, 'r+');
+
+		// we need to acquire a lock to avoid multiple migrations at the same
+		// time if several users try to access the application.
+		if (!flock($applied_migrations_file, LOCK_EX)) {
```

Comment:
I am a bit worried about this part. I have not had the energy to test on my side, but we should ensure that it works (at least in theory) with different Web servers, OS, mounts (e.g. network drives...), especially when used at such a critical place.
Acquiring a lock is also expensive and error prone (at least last time I checked), and this worries me for scaling (so a two-step approach should be considered: testing a cheaper condition such as file existence, and then go for the lock).
We should also release the lock fast, even when the request is aborted (with the current code, I have not checked but we might end up in an indeterminate state - see e.g. https://php.net/function.ignore-user-abort ).
Finally, it would be good to have a fail safe approach, i.e. that we anyway try to run FreshRSS instead of returning an error message.
Many years ago, I did a few experiments with different methods, still visible in https://alexandre.alapetite.fr/doc-alex/compteur/ and it was not trivial to get right.

Reply:
P.S. A database request can also be considered

Reply:
You're right, this is not trivial and plus, I didn't think about a potential user abort. Retrospectively I wonder if it shouldn't be better to force the admin to apply the migration himself (via the CLI, or during the automatic update) so we can simply get rid of this lock-approach. What do you think?

>  Finally, it would be good to have a fail safe approach, i.e. that we anyway try to run FreshRSS instead of returning an error message.

I would rather show a maintenance page to normal users, and explain to administrators how to apply the update. I don't like the idea of bypassing an error because it's often the cause of more errors. If the update doesn't apply, it's almost certain something else will not work then, and FRSS will look broken.

Reply:
I changed the system in https://github.com/FreshRSS/FreshRSS/pull/2760/commits/8e05d9115cac334790b0d1f133785a15879e6bfc but it brings different problems. I'm afraid that there's no perfect system if we want to let the migrations being applied on user requests.

The more robust would be to apply them on an update and show a maintenance page while there are not.

Reply:
I do not have much better solution to propose right now, and as long as it does not introduce a potential bottleneck or issue with some platforms, I think that would be fine.